File: src/main/java/me/ag2s/epublib/domain/TableOfContents.java
package me.ag2s.epublib.domain;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.Collection;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class TableOfContents implements Serializable {
   private static final long serialVersionUID = -3147391239966275152L;
   public static final String DEFAULT_PATH_SEPARATOR = "/";
   private List tocReferences;

   public TableOfContents() {
      this(new ArrayList());
   }

   public TableOfContents(List tocReferences) {
      this.tocReferences = tocReferences;
   }

   public List getTocReferences() {
      return this.tocReferences;
   }

   public void setTocReferences(List tocReferences) {
      this.tocReferences = tocReferences;
   }

   public TOCReference addSection(Resource resource, String path) {
      return this.addSection(resource, path, "/");
   }

   public TOCReference addSection(Resource resource, String path, String pathSeparator) {
      String[] pathElements = path.split(pathSeparator);
      return this.addSection(resource, pathElements);
   }

   private static TOCReference findTocReferenceByTitle(String title, List tocReferences) {
      for(TOCReference tocReference : tocReferences) {
         if (title.equals(tocReference.getTitle())) {
            return tocReference;
         }
      }

      return null;
   }

   public TOCReference addSection(Resource resource, String[] pathElements) {
      if (pathElements != null && pathElements.length != 0) {
         TOCReference result = null;
         List<TOCReference> currentTocReferences = this.tocReferences;

         for(String currentTitle : pathElements) {
            result = findTocReferenceByTitle(currentTitle, currentTocReferences);
            if (result == null) {
               result = new TOCReference(currentTitle, (Resource)null);
               currentTocReferences.add(result);
            }

            currentTocReferences = result.getChildren();
         }

         result.setResource(resource);
         return result;
      } else {
         return null;
      }
   }

   public TOCReference addSection(Resource resource, int[] pathElements, String sectionTitlePrefix, String sectionNumberSeparator) {
      if (pathElements != null && pathElements.length != 0) {
         TOCReference result = null;
         List<TOCReference> currentTocReferences = this.tocReferences;

         for(int i = 0; i < pathElements.length; ++i) {
            int currentIndex = pathElements[i];
            if (currentIndex > 0 && currentIndex < currentTocReferences.size() - 1) {
               result = (TOCReference)currentTocReferences.get(currentIndex);
            } else {
               result = null;
            }

            if (result == null) {
               this.paddTOCReferences(currentTocReferences, pathElements, i, sectionTitlePrefix, sectionNumberSeparator);
               result = (TOCReference)currentTocReferences.get(currentIndex);
            }

            currentTocReferences = result.getChildren();
         }

         result.setResource(resource);
         return result;
      } else {
         return null;
      }
   }

   private void paddTOCReferences(List currentTocReferences, int[] pathElements, int pathPos, String sectionPrefix, String sectionNumberSeparator) {
      for(int i = currentTocReferences.size(); i <= pathElements[pathPos]; ++i) {
         String sectionTitle = this.createSectionTitle(pathElements, pathPos, i, sectionPrefix, sectionNumberSeparator);
         currentTocReferences.add(new TOCReference(sectionTitle, (Resource)null));
      }

   }

   private String createSectionTitle(int[] pathElements, int pathPos, int lastPos, String sectionPrefix, String sectionNumberSeparator) {
      StringBuilder title = new StringBuilder(sectionPrefix);

      for(int i = 0; i < pathPos; ++i) {
         if (i > 0) {
            title.append(sectionNumberSeparator);
         }

         title.append(pathElements[i] + 1);
      }

      if (pathPos > 0) {
         title.append(sectionNumberSeparator);
      }

      title.append(lastPos + 1);
      return title.toString();
   }

   public TOCReference addTOCReference(TOCReference tocReference) {
      if (this.tocReferences == null) {
         this.tocReferences = new ArrayList();
      }

      this.tocReferences.add(tocReference);
      return tocReference;
   }

   public List getAllUniqueResources() {
      Set<String> uniqueHrefs = new HashSet();
      List<Resource> result = new ArrayList();
      getAllUniqueResources(uniqueHrefs, result, this.tocReferences);
      return result;
   }

   private static void getAllUniqueResources(Set uniqueHrefs, List result, List tocReferences) {
      for(TOCReference tocReference : tocReferences) {
         Resource resource = tocReference.getResource();
         if (resource != null && !uniqueHrefs.contains(resource.getHref())) {
            uniqueHrefs.add(resource.getHref());
            result.add(resource);
         }

         getAllUniqueResources(uniqueHrefs, result, tocReference.getChildren());
      }

   }

   public int size() {
      return getTotalSize(this.tocReferences);
   }

   private static int getTotalSize(Collection tocReferences) {
      int result = tocReferences.size();

      for(TOCReference tocReference : tocReferences) {
         result += getTotalSize(tocReference.getChildren());
      }

      return result;
   }

   public int calculateDepth() {
      return this.calculateDepth(this.tocReferences, 0);
   }

   private int calculateDepth(List tocReferences, int currentDepth) {
      int maxChildDepth = 0;

      for(TOCReference tocReference : tocReferences) {
         int childDepth = this.calculateDepth(tocReference.getChildren(), 1);
         if (childDepth > maxChildDepth) {
            maxChildDepth = childDepth;
         }
      }

      return currentDepth + maxChildDepth;
   }
}

File: src/main/kotlin/me/ag2s/epublib/domain/TableOfContents.kt
package me.ag2s.epublib.domain

import java.io.Serializable

class TableOfContents : Serializable {
    private var tocReferences: MutableList<TOCReference>
    constructor() : this(ArrayList())
    constructor(tocReferences: MutableList<TOCReference>) { this.tocReferences = tocReferences }
    fun getTocReferences(): MutableList<TOCReference> = tocReferences
    fun setTocReferences(tocReferences: MutableList<TOCReference>) { this.tocReferences = tocReferences }
    fun addSection(resource: Resource?, path: String): TOCReference? = addSection(resource, path, DEFAULT_PATH_SEPARATOR)
    fun addSection(resource: Resource?, path: String, pathSeparator: String): TOCReference? = addSection(resource, path.split(pathSeparator).toTypedArray())
    fun addSection(resource: Resource?, pathElements: Array<String>?): TOCReference? { if (pathElements.isNullOrEmpty()) return null; var current = tocReferences; var result: TOCReference? = null; pathElements.forEach { title -> result = current.firstOrNull { it.getTitle() == title } ?: TOCReference(title, null).also { current.add(it) }; current = result!!.getChildren() }; result!!.setResource(resource); return result }
    fun addSection(resource: Resource?, pathElements: IntArray?, sectionTitlePrefix: String, sectionNumberSeparator: String): TOCReference? {
        if (pathElements == null || pathElements.isEmpty()) return null
        var current = tocReferences
        var result: TOCReference? = null
        pathElements.forEachIndexed { position, currentIndex ->
            result = if (currentIndex > 0 && currentIndex < current.size - 1) current[currentIndex] else null
            if (result == null) {
                for (index in current.size..currentIndex) current.add(TOCReference(createSectionTitle(pathElements, position, index, sectionTitlePrefix, sectionNumberSeparator), null))
                result = current[currentIndex]
            }
            current = result!!.getChildren()
        }
        result!!.setResource(resource)
        return result
    }
    private fun createSectionTitle(pathElements: IntArray, pathPosition: Int, lastPosition: Int, prefix: String, separator: String): String = buildString { append(prefix); for (index in 0 until pathPosition) { if (index > 0) append(separator); append(pathElements[index] + 1) }; if (pathPosition > 0) append(separator); append(lastPosition + 1) }
    fun addTOCReference(reference: TOCReference): TOCReference { tocReferences.add(reference); return reference }
    fun getAllUniqueResources(): MutableList<Resource> { val hrefs = HashSet<String?>(); val result = ArrayList<Resource>(); fun add(references: List<TOCReference>) { references.forEach { reference -> reference.getResource()?.let { resource -> if (hrefs.add(resource.getHref())) result.add(resource) }; add(reference.getChildren()) } }; add(tocReferences); return result }
    fun size(): Int {
        fun count(references: Collection<TOCReference>): Int {
            var result = references.size
            references.forEach { result += count(it.getChildren()) }
            return result
        }
        return count(tocReferences)
    }
    fun calculateDepth(): Int { fun depth(references: List<TOCReference>, current: Int): Int { var maximum = 0; references.forEach { maximum = maxOf(maximum, depth(it.getChildren(), 1)) }; return current + maximum }; return depth(tocReferences, 0) }
    companion object { private const val serialVersionUID = -3147391239966275152L; const val DEFAULT_PATH_SEPARATOR = "/" }
}
